### ISSUE TITLE
stubtest: Support `type_check_only` decorator

### ISSUE THREAD

**Issue by Avasam** [python/mypy#15146]
**Feature**

I would like stubtest to support checking the `type_check_only` decorator.
This means something marked with `type_check_only` that DOES exist at runtime is an error.
And something that DOES NOT exist at runtime, and is marked with `type_check_only` should pass.

**Pitch**

Instead of marking stub-only classes and methods as private in typeshed, they can be marked with `@type_check_only`, which is designed for that exact purpose. Private may or may not be found at runtime, and users sometimes don't have a choice to use them, and basically have to guess whether they need to wrap the import in `TYPE_CHECK_ONLY` or not.
This would also promote marking as special methods that don't exist at runtime, but are added to support some typechecker features, in typeshed.

Related: https://github.com/python/mypy/issues/9531

**Comment by Avasam** [comment]
Could I get the `topic-stubtest` label here?

**Comment by jorenham** [comment]
There is partial support for this (at least with `>=1.15.0`):

https://github.com/python/mypy/blob/16e99de5376464beaa2cf086c1cd3dc5d26a791a/mypy/stubtest.py#L508-L513

https://github.com/python/mypy/blob/16e99de5376464beaa2cf086c1cd3dc5d26a791a/mypy/stubtest.py#L1378-L1383

But in case of overloaded `@type_check_only` functions, stubtest will still report a false positive "error: ... is not present at runtime":

https://github.com/python/mypy/blob/16e99de5376464beaa2cf086c1cd3dc5d26a791a/mypy/stubtest.py#L1163-L1166

I just ran into this limitation [scipy-stubs](https://github.com/scipy/scipy-stubs), so it would be nice if it would also work for overload functions.

**Comment by hauntsaninja** [comment]
Yeah, looks like sobolevn left this as a TODO in https://github.com/python/mypy/pull/16422

To answer the question he left, I'd assume the semantics are that it applies to the whole function, is that what you expect as well?

**Comment by jorenham** [comment]
> I'd assume the semantics are that it applies to the whole function, is that what you expect as well?

I can't imagine where you'd want to `@type_check_only` a single overload 🤔. So yea, I think it would make sense to have one `@type_check_only` apply to the entire overloaded method. So like `@override` and `final`, and not like `@deprecated`.
